File: StepFunctionInvoke/typescript/lib/pipeline-stack.ts
import * as cdk from 'aws-cdk-lib';
import * as codebuild from 'aws-cdk-lib/aws-codebuild';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as stepfunctions from 'aws-cdk-lib/aws-stepfunctions';
import { Construct } from 'constructs';
import { CodePipeline, CodePipelineSource, ShellStep, ManualApprovalStep } from 'aws-cdk-lib/pipelines';
import { AppStage } from './app-stage';

/**
 * Pipeline stack with Step Function invocation for complex workflow orchestration.
 * Includes automated validation and workflow management after each deployment stage.
 */
export class PipelineStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // Create Step Function for deployment validation
    const validateTask = new stepfunctions.Pass(this, 'ValidateDeployment', {
      comment: 'Validate deployment resources',
      parameters: {
        'ValidationResult': 'SUCCESS',
        'Timestamp.$': '$$.State.EnteredTime',
        'Input.$': '$'
      }
    });

    const checkHealthTask = new stepfunctions.Pass(this, 'CheckHealth', {
      comment: 'Check application health',
      parameters: {
        'HealthStatus': 'HEALTHY',
        'CheckedAt.$': '$$.State.EnteredTime'
      }
    });

    const notifyTask = new stepfunctions.Pass(this, 'NotifySuccess', {
      comment: 'Notify successful validation',
      parameters: {
        'Message': 'Deployment validation completed successfully',
        'CompletedAt.$': '$$.State.EnteredTime'
      }
    });

    const definition = validateTask
      .next(checkHealthTask)
      .next(notifyTask)
      .next(new stepfunctions.Succeed(this, 'ValidationComplete'));

    const validationStateMachine = new stepfunctions.StateMachine(this, 'ValidationStateMachine', {
      definition,
      timeout: cdk.Duration.minutes(10),
    });

    const pipeline = new CodePipeline(this, 'Pipeline', {
      pipelineName: 'StepFunctionInvokePipeline',
      synth: new ShellStep('Synth', {
        input: CodePipelineSource.gitHub('OWNER/REPO', 'main', {
          authentication: cdk.SecretValue.secretsManager('github-token')
        }),
        commands: [
          'npm ci',
          'npm run build',
          'npx cdk synth'
        ]
      }),
      dockerEnabledForSynth: true,
      dockerEnabledForSelfMutation: true,
      synthCodeBuildDefaults: {
        buildEnvironment: {
          buildImage: codebuild.LinuxBuildImage.STANDARD_7_0
        },
        rolePolicy: [
          new iam.PolicyStatement({
            effect: iam.Effect.ALLOW,
            actions: [
              'ec2:DescribeAvailabilityZones',
              'ec2:DescribeVpcs',
              'ec2:DescribeSubnets',
              'ec2:DescribeRouteTables',
              'ec2:DescribeSecurityGroups',
              'ssm:GetParameter',
              'ssm:GetParameters'
            ],
            resources: ['*']
          })
        ]
      },
      selfMutationCodeBuildDefaults: {
        buildEnvironment: {
          buildImage: codebuild.LinuxBuildImage.STANDARD_7_0
        }
      }
    });

    // Add test stage
    const testStage = pipeline.addStage(new AppStage(this, 'Test', {
      env: { 
        account: process.env.CDK_DEFAULT_ACCOUNT, 
        region: process.env.CDK_DEFAULT_REGION 
      }
    }));

    testStage.addPre(new ShellStep('UnitTest', {
      commands: [
        'npm ci',
        'npm test'
      ]
    }));

    // Add Step Function invocation after test deployment
    testStage.addPost(new ShellStep('InvokeStepFunction', {
      commands: [
        `EXECUTION_ARN=$(aws stepfunctions start-execution --state-machine-arn ${validationStateMachine.stateMachineArn} --input '{"stage":"test","deploymentId":"'$(date +%s)'"}' --query 'executionArn' --output text)`,
        'echo "Started Step Function execution: $EXECUTION_ARN"',
        'aws stepfunctions describe-execution --execution-arn $EXECUTION_ARN',
        'echo "Step Function validation initiated successfully"'
      ]
    }));

    // Add production stage with manual approval
    const prodStage = pipeline.addStage(new AppStage(this, 'Prod', {
      env: { 
        account: process.env.CDK_DEFAULT_ACCOUNT, 
        region: process.env.CDK_DEFAULT_REGION 
      }
    }));

    prodStage.addPre(new ManualApprovalStep('PromoteToProd'));

    // Add Step Function invocation after production deployment
    prodStage.addPost(new ShellStep('InvokeStepFunction', {
      commands: [
        `EXECUTION_ARN=$(aws stepfunctions start-execution --state-machine-arn ${validationStateMachine.stateMachineArn} --input '{"stage":"prod","deploymentId":"'$(date +%s)'"}' --query 'executionArn' --output text)`,
        'echo "Started Step Function execution: $EXECUTION_ARN"',
        'aws stepfunctions describe-execution --execution-arn $EXECUTION_ARN',
        'echo "Step Function validation initiated successfully"'
      ]
    }));

    // Outputs
    new cdk.CfnOutput(this, 'StateMachineArn', {
      value: validationStateMachine.stateMachineArn,
      description: 'Step Function State Machine ARN for validation',
    });

    new cdk.CfnOutput(this, 'StateMachineName', {
      value: validationStateMachine.stateMachineName,
      description: 'Step Function State Machine Name',
    });
  }
}

File: ECRBuild/python/ecrbuild/pipeline_stack.py
"""
Pipeline stack with multi-stage deployment.
"""
import aws_cdk as cdk
from aws_cdk import pipelines, aws_iam as iam, aws_codebuild as codebuild
from constructs import Construct
from .app_stack import AppStack

class AppStage(cdk.Stage):
    def __init__(self, scope: Construct, construct_id: str, **kwargs) -> None:
        super().__init__(scope, construct_id, **kwargs)
        AppStack(self, "AppStack")

class PipelineStack(cdk.Stack):
    def __init__(self, scope: Construct, construct_id: str, **kwargs) -> None:
        super().__init__(scope, construct_id, **kwargs)

        pipeline = pipelines.CodePipeline(
            self, "Pipeline",
            pipeline_name="ECRBuildPipeline",
            synth=pipelines.ShellStep(
                "Synth",
                input=pipelines.CodePipelineSource.git_hub(
                    "OWNER/REPO", "main",
                    authentication=cdk.SecretValue.secrets_manager("github-token")
                ),
                commands=[
                    "pip install -r requirements.txt",
                    "npx cdk synth"
                ]
            ,
            synth_code_build_defaults=pipelines.CodeBuildOptions(
                build_environment=codebuild.BuildEnvironment(
                    build_image=codebuild.LinuxBuildImage.STANDARD_7_0
                ),
                role_policy=[
                    iam.PolicyStatement(
                        effect=iam.Effect.ALLOW,
                        actions=[
                            "ec2:DescribeAvailabilityZones",
                            "ec2:DescribeVpcs",
                            "ec2:DescribeSubnets",
                            "ec2:DescribeRouteTables",
                            "ec2:DescribeSecurityGroups",
                            "ssm:GetParameter",
                            "ssm:GetParameters"
                        ],
                        resources=["*"]
                    )
                ]
            ),
            self_mutation_code_build_defaults=pipelines.CodeBuildOptions(
                build_environment=codebuild.BuildEnvironment(
                    build_image=codebuild.LinuxBuildImage.STANDARD_7_0
                )
            )
        )
        ,
            synth_code_build_defaults=pipelines.CodeBuildOptions(
                build_environment=codebuild.BuildEnvironment(
                    build_image=codebuild.LinuxBuildImage.STANDARD_7_0
                ),
                role_policy=[
                    iam.PolicyStatement(
                        effect=iam.Effect.ALLOW,
                        actions=[
                            "ec2:DescribeAvailabilityZones",
                            "ec2:DescribeVpcs",
                            "ec2:DescribeSubnets",
                            "ec2:DescribeRouteTables",
                            "ec2:DescribeSecurityGroups",
                            "ssm:GetParameter",
                            "ssm:GetParameters"
                        ],
                        resources=["*"]
                    )
                ]
            ),
            self_mutation_code_build_defaults=pipelines.CodeBuildOptions(
                build_environment=codebuild.BuildEnvironment(
                    build_image=codebuild.LinuxBuildImage.STANDARD_7_0
                )
            )
        )

        test_stage = pipeline.add_stage(AppStage(self, "Test", env=cdk.Environment(
            account=self.account, region=self.region
        )))

        test_stage.add_pre(pipelines.ShellStep("UnitTest", commands=[
            "pip install -r requirements.txt",
            "python -m pytest tests/ -v"
        ]))

        prod_stage = pipeline.add_stage(AppStage(self, "Prod", env=cdk.Environment(
            account=self.account, region=self.region
        )))

        prod_stage.add_pre(pipelines.ManualApprovalStep("PromoteToProd"))
